### PR TITLE
Fix boundary conditions

### PR DESCRIPTION
**Pull request by Homingdung**
To allow `DirichletBC(V, 0, ...) `, the special case requires no computation for stage boundary condition, everything should be zero

### CODE REVIEW

**Review by rckirby — CHANGES_REQUESTED** [review]
Thanks for tackling this, it's long been a major annoyance.  However, I tried out your change elsewhere (in `test_stokes.py`) and find that that is somehow broken.  I expected stage value formulation to fail in that test since we haven't done similar updates to how that handles BC, but the classic stage-derivative form you attempted to fix is failing here.  Please take a look.

File: irksome/stage.py [hunk 1]
```diff
@@ -50,6 +50,20 @@ def getBits(num_stages, num_fields, u0, UU, v, VV):
     return u0bits, vbits, VVbits, UUbits
 
 
+def gstuff(V, g):
+    if g == 0:
+        gdat = 0
```
Comment (pbrubeck) [inline irksome/stage.py:55]:
```suggestion
        gdat = g
```

File: irksome/getForm.py [hunk 2]
```diff
@@ -15,41 +15,33 @@
 
 class BCStageData(object):
     def __init__(self, V, gcur, u0, u0_mult, i, t, dt):
-        if V.component is not None:     # bottommost space is bit of VFS
+        if gcur == 0:  # special case DirichletBC(V, 0, ...), do nothing
+            u0 = 0
```
Comment (pbrubeck) [inline irksome/getForm.py:19]:
```suggestion
        if gcur == 0:  # special case DirichletBC(V, 0, ...), do nothing
            pass
```

File: irksome/getForm.py [hunk 3]
```diff
@@ -15,41 +15,33 @@
...
-                try:
-                    gdat = assemble(interpolate(gcur-u0_mult[i]*u0.sub(sub), V))
-                    gmethod = lambda g, u: gdat.interpolate(g-u0_mult[i]*u.sub(sub))
-                except:  # noqa: E722
-                    gdat = project(gcur-u0_mult[i]*u0.sub(sub), V)
-                    gmethod = lambda g, u: gdat.project(g-u0_mult[i]*u.sub(sub))
-
+                u0 = u0.sub(sub)
+        if u0 == 0:
+            gdat = gcur
```
Comment (pbrubeck) [inline irksome/getForm.py:35]:
```suggestion
        if gcur == 0:
            gdat = gcur
```

File: irksome/getForm.py [hunk 4]
```diff
@@ -15,41 +15,33 @@
...
+            gmethod = lambda g, u: None
+        else:
+            gexpr = gcur-u0_mult[i]*u0
+            try:
+                gdat = assemble(interpolate(gexpr, V))
+                gmethod = lambda g, u: gdat.interpolate(gexpr)
+            except (NotImplementedError, AttributeError):
+                gdat = project(gexpr, V)
+                gmethod = lambda g, u: gdat.project(gexpr)
```
Comment (pbrubeck) [inline irksome/getForm.py:44]: Are both `g` and `u` changing each time?

Comment (rckirby) [reply irksome/getForm.py]: In principle, they could be.  Looking at this now after a few years, it seems that `g` is unused in the lambda expression...

File: irksome/stage.py [hunk 5]
```diff
@@ -271,44 +260,26 @@ def getFormStage(F, butch, u0, t, dt, bcs=None, splitting=None, vandermonde=None
...
             bcparams, bclower, bcupper = bc_constraints[bc]
-            for i in range(num_stages):
-                Vsp, Vbigi = stage2spaces4bc(bc, V, Vbig, i)
-                gdat = Function(Vsp)
-                vbc = TestFunction(Vsp)
-                gmethod = lambda gd, gc: solve(inner(gd - gc, vbc) * dx == 0,
-                                               gd, solver_parameters=bcparams)
             gcur = replace(bcarg, {t: t+C[i] * dt})
-            gblah_cur.append((gdat, gcur - Vander_col[i] * bcarg, gmethod))
+            gcur = gcur - Vander_col[i] * bcarg
```
Comment (rckirby) [inline irksome/stage.py:266]: Is there a reason to prefer this over `gcur -= ...`?

File: irksome/stage.py [hunk 6]
```diff
@@ -271,44 +260,26 @@ def getFormStage(F, butch, u0, t, dt, bcs=None, splitting=None, vandermonde=None
...
-
-                gcur = replace(bcarg, {t: t+C[i]*dt})
-                gblah_cur.append((gdat, gcur - Vander_col[i] * bcarg, gmethod))
-
             gdats_cur = np.zeros((num_stages,), dtype="O")
             for i in range(num_stages):
-                gdats_cur[i] = gblah_cur[i][0]
+                Vbigi = stage2spaces4bc(bc, V, Vbig, i)
+                gcur = replace(bcarg, {t: t+C[i]*dt})
+                gcur = gcur - Vander_col[i] * bcarg
```
Comment (rckirby) [inline irksome/stage.py:272]: same here.

File: irksome/stage.py [hunk 7]
```diff
@@ -271,44 +260,26 @@ def getFormStage(F, butch, u0, t, dt, bcs=None, splitting=None, vandermonde=None
     for bc in bcs:
         bcarg = as_ufl(bc._original_arg)
 
-        gblah_cur = []
-
         if bc in bc_constraints:
             bcparams, bclower, bcupper = bc_constraints[bc]
```
Comment (rckirby) [inline irksome/stage.py:264]: `bcparams`, `bclower`, etc seem unused now?  The idea of this bit is that we need to be able to pass in special parameters to project the boundary condition expression into the finite element space subject to bounds constraints (e.g. using a VI solver).  It looks like this capability just disappeared.

**Review by rckirby — CHANGES_REQUESTED** [review]
I think this is very very clean.  Except we have a technical corner case about no longer being able to do bounds-constrained projections for discrete boundary conditions.

File: irksome/stage.py (hunk 5 above)
Comment (pbrubeck) [reply irksome/stage.py:266]: Constants cannot be updated

**Review by ScottMacLachlan — COMMENTED** [review]
Modulo answers to @rckirby 's questions about the bounds-constrained case, this all makes sense to me.  Seems much cleaner than what we were doing, and resolves a few things as well!  Many thanks!

**Comment by rckirby** [comment]
Right now, our only use cases are constant upper and/or lower bounds.  I
think we should not make the perfect the enemy of the good.  For now, if
you have a `Function` in the upper or lower bound and update it before the
solver is called, the new bounds will be in place.  We don't have a
mechanism for `bounds` being an expression, but that's not an issue because
1) FD's bounds mechanism is the same and 2) you'd need a bounds-constrained
projection to pull it off..

File: irksome/stage.py (hunk 5 above)
Comment (rckirby) [reply irksome/stage.py:266]: Ah, I didn't think about that case.

**Review by rckirby — APPROVED** [review]

**Pull request merged**